### PR TITLE
drivers: Add driver to disable flash patch

### PR DESCRIPTION
**Pull request by oyvindronningstad**
If enabled, write to UICR the first time bootloader boots.

Signed-off-by: Øyvind Rønningstad <oyvind.ronningstad@nordicsemi.no>

### CODE REVIEW

**Comment by oyvindronningstad** [comment]
cc @sigvartmh @frkv @osaether

File: subsys/bootloader/Kconfig [hunk 1]
```diff
@@ -154,6 +154,15 @@ config SB_BPROT_IN_DEBUG
 	depends on (HAS_HW_NRF_BPROT || HAS_HW_NRF_MPU)
 
 endif # SECURE_BOOT_DEBUG
+
+config SB_DISABLE_FLASH_PATCH
+	bool "Disable Cortex-M4 Flash Patch capabilities"
+	depends on SOC_NRF52840
+	help
+	  The flash patch can be used by malicious code to circumvent secure
```
Comment (hakonfam) [inline subsys/bootloader/Kconfig]: IMO the commit message should contain this information.

File: samples/bootloader/src/main.c [hunk 2]
```diff
@@ -178,6 +201,14 @@ void main(void)
 		return;
 	}
 
+#ifdef CONFIG_SB_DISABLE_FLASH_PATCH
+	if ((NRF_UICR->DEBUGCTRL & UICR_DEBUGCTRL_CPUFPBEN_Msk) != 0x0) {
```
Comment (hakonfam) [inline samples/bootloader/src/main.c]: Some comment here on what this does and why would be good.

File: samples/bootloader/src/main.c [hunk 3]
```diff
@@ -26,6 +26,29 @@
 #endif
 #endif
 
+
+#ifdef CONFIG_SB_DISABLE_FLASH_PATCH
+/* Copied from nrf_nvmc.c */
```
Comment (hakonfam) [inline samples/bootloader/src/main.c]: Do we need to have this here? IMO it does not matter where it comes from.

File: samples/bootloader/src/main.c [hunk 4]
```diff
@@ -178,6 +201,14 @@ void main(void)
 		return;
 	}
 
+#ifdef CONFIG_SB_DISABLE_FLASH_PATCH
+	if ((NRF_UICR->DEBUGCTRL & UICR_DEBUGCTRL_CPUFPBEN_Msk) != 0x0) {
+		nrf_nvmc_write_word((uint32_t)&NRF_UICR->DEBUGCTRL,
```
Comment (hakonfam) [inline samples/bootloader/src/main.c]: If we are not planning on using the `nrf_nvmc_write_word` function for anything else, could we not put all of this inside a function and have the function name work as documentation?

E.g. `disable_debug_control()` and then describe inside what it does. Also we can get rid of the `#ifdef` inside main()

Comment (oyvindronningstad) [reply samples/bootloader/src/main.c]: I'll move the `if ((NRF_UICR...` into the function and rename it, that's cleaner I agree. I'd rather keep the ifdef in main though, to show that it doesn't happen always.

Comment (hakonfam) [reply samples/bootloader/src/main.c]: Good point, I agree.

File: subsys/bootloader/Kconfig (hunk 1 above)
Comment (oyvindronningstad) [reply subsys/bootloader/Kconfig]: Done

File: samples/bootloader/src/main.c (hunk 2 above)
Comment (oyvindronningstad) [reply samples/bootloader/src/main.c]: Added a comment (in the new function)

File: samples/bootloader/src/main.c (hunk 3 above)
Comment (oyvindronningstad) [reply samples/bootloader/src/main.c]: Removed

File: samples/bootloader/src/main.c (hunk 4 above)
Comment (oyvindronningstad) [reply samples/bootloader/src/main.c]: Done

File: samples/bootloader/src/main.c [hunk 5]
```diff
@@ -25,6 +25,35 @@
...
 
 
+#ifdef CONFIG_SB_DISABLE_FLASH_PATCH
+static void disable_flash_patch(void)
+{
+	/* Check if register has been written. */
+	if ((NRF_UICR->DEBUGCTRL & UICR_DEBUGCTRL_CPUFPBEN_Msk) != 0x0) {
+		/* Enable flash write. */
+		NRF_NVMC->CONFIG = NVMC_CONFIG_WEN_Wen;
+		__ISB();
```
Comment (ioannisg) [inline samples/bootloader/src/main.c]: Reverse the order of DSB and ISB

File: samples/bootloader/src/main.c [hunk 6]
```diff
@@ -25,6 +25,35 @@
...
+		/* Write to register. */
+		NRF_UICR->DEBUGCTRL = ~UICR_DEBUGCTRL_CPUFPBEN_Msk;
+
+		/* Wait for flash ready */
+		while (NRF_NVMC->READY == NVMC_READY_READY_Busy)
+			;
+
+		/* Disable flash write. */
+		NRF_NVMC->CONFIG = NVMC_CONFIG_WEN_Ren;
+		__ISB();
```
Comment (ioannisg) [inline samples/bootloader/src/main.c]: Reverse the order of ISB DSB

File: samples/bootloader/src/main.c [hunk 7]
```diff
@@ -25,6 +25,35 @@
...
+		__DSB();
+
+		/* Reset for change to take effect. */
+		NVIC_SystemReset();
```
Comment (ioannisg) [inline samples/bootloader/src/main.c]: IS it possible to call the zephyr reset api function?

File: samples/bootloader/src/main.c (hunk 5 above)
Comment (oyvindronningstad) [reply samples/bootloader/src/main.c]: Done

File: samples/bootloader/src/main.c (hunk 6 above)
Comment (oyvindronningstad) [reply samples/bootloader/src/main.c]: Done

File: samples/bootloader/src/main.c (hunk 7 above)
Comment (oyvindronningstad) [reply samples/bootloader/src/main.c]: Done

**Comment by sigvartmh** [comment]
@hakonfam or @ioannisg I think this can be merged

**Comment by hakonfam** [comment]
One last comment: Should we have some sort of  "THIS MUST BE SET IN PRODUCTION" warning? Like the one there is for ASSERT?

I was thinking of checking the config in cmake an print a huge error message there. That way you only get it when doing the initial configuration, and not compiling.

**Comment by sigvartmh** [comment]
`#warning` in compiler and output in cmake?

**Comment by oyvindronningstad** [comment]
> One last comment: Should we have some sort of "THIS MUST BE SET IN PRODUCTION" warning? Like the one there is for ASSERT?
> 
> I was thinking of checking the config in cmake an print a huge error message there. That way you only get it when doing the initial configuration, and not compiling.

Added a warning like for ASSERTs

**Comment by oyvindronningstad** [comment]
Should we have this in MCUboot as well?

**Comment by hakonfam** [comment]
> Should we have this in MCUboot as well?

I think so, this should be upstream AFAICT.

**Comment by sigvartmh** [comment]
It will have to be nRF specific so I guess upstream to zephyr?

File: tests/subsys/bootloader/flash_patch/test.sh [hunk 8]
```diff
@@ -0,0 +1,27 @@
+#!/bin/bash
```
Comment (hakonfam) [inline tests/subsys/bootloader/flash_patch/test.sh]: Is this script pointed to somewhere? If not, I fear for code rot. Either it should be executed by CI, or we should point to it in some documentation.

Comment (hakonfam) [reply tests/subsys/bootloader/flash_patch/test.sh]: ping @oyvindronningstad , would be good to clarify this before merging.

Comment (oyvindronningstad) [reply tests/subsys/bootloader/flash_patch/test.sh]: I'll update docs

Comment (oyvindronningstad) [reply tests/subsys/bootloader/flash_patch/test.sh]: Turns out we don't have detailed docs for the immutable bootloader, so I don't know where to put this.

Comment (oyvindronningstad) [reply tests/subsys/bootloader/flash_patch/test.sh]: This will be hooked up to CI.

**Comment by oyvindronningstad** [comment]
> > Should we have this in MCUboot as well?
> 
> I think so, this should be upstream AFAICT.

I might try to make this into a driver.

**Comment by oyvindronningstad** [comment]
> > > Should we have this in MCUboot as well?
> > 
> > 
> > I think so, this should be upstream AFAICT.
> 
> I might try to make this into a driver.

Done. It's not available in both MCUboot and B0, and gives warning in both if not present.

File: drivers/flash_patch/flash_patch.c [hunk 9]
```diff
@@ -0,0 +1,42 @@
+/*
+ * Copyright (c) 2018 Nordic Semiconductor ASA
```
Comment (hakonfam) [inline drivers/flash_patch/flash_patch.c]: 2019

File: drivers/flash_patch/Kconfig [hunk 10]
```diff
@@ -0,0 +1,9 @@
+
```
Comment (hakonfam) [inline drivers/flash_patch/Kconfig]: Missing copyright?

File: drivers/CMakeLists.txt [hunk 11]
```diff
@@ -13,3 +13,4 @@ add_subdirectory_ifdef(CONFIG_ADP536X adp536x)
 add_subdirectory_ifdef(CONFIG_ST25R3911B_LIB st25r3911b)
 add_subdirectory_ifdef(CONFIG_BT_LL_NRFXLIB bt_ll_nrfxlib)
 add_subdirectory_ifdef(CONFIG_NRF9160_GPS nrf9160_gps)
+add_subdirectory(flash_patch)
```
Comment (hakonfam) [inline drivers/CMakeLists.txt]: I assume it is correct that this should always included.

However, is there any point in including it when no module which locks the flash is being used?

File: drivers/flash_patch/CMakeLists.txt [hunk 12]
```diff
@@ -0,0 +1,14 @@
+
```
Comment (hakonfam) [inline drivers/flash_patch/CMakeLists.txt]: Missing copyright

File: drivers/CMakeLists.txt (hunk 11 above)
Comment (hakonfam) [reply drivers/CMakeLists.txt]: I see now that the checks are all performed in the `flash_patch` cmakelist, so I guess you can disregard this comment.

Comment (oyvindronningstad) [reply drivers/CMakeLists.txt]: It's done this way because of the warning test.

**Comment by oyvindronningstad** [comment]
Thanks, @hakonfam, copyrights are fixed now.

**Comment by oyvindronningstad** [comment]
> Looks good. Remember that this won't be merged until the changelog has been updated with the change. Also remember to create the doc page referring to the test. Alternatively create a task for running it during CI.

Updated changelog and created task.

**Pull request merged**